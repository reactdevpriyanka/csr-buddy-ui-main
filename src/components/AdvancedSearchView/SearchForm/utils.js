import { subWeeks, isValid } from 'date-fns';
import * as Yup from 'yup';

export const phoneRegExp = /^(\+?\d{0,4})?\s?-?\s?(\(?\d{3}\)?)\s?-?\s?(\(?\d{3}\)?)\s?-?\s?(\(?\d{4}\)?)?$/;

export const validationSchema = Yup.object({
  orderId: Yup.string().trim().matches(/^\d*$/, 'Must be numeric characters'),
  status: Yup.string().trim(),
  name: Yup.string().trim(),
  logonId: Yup.string().trim(),
  address: Yup.string().trim(),
  city: Yup.string().trim(),
  zip: Yup.string().matches(
    /^\d{5}(?:-\d{4})?$/,
    'Must be 5 or 9 digits with a format of ##### or #####-####',
  ),
  email: Yup.string().email().trim(),
  phone: Yup.string()
    .trim()
    .matches(phoneRegExp, 'Phone number is not valid')
    .min(7, 'Phone number is not valid')
    .max(16, 'Phone number is not valid'),
  partNumber: Yup.string().trim(),
  blocked: Yup.string().trim(),
  blockReason: Yup.string().trim(),
  fulfillmentCenter: Yup.string().trim(),
  account: Yup.string().trim().min(4).max(4),
  payReferenceId: Yup.string().trim(),
  paypalEmail: Yup.string().email().trim(),
  ipAddress: Yup.string().trim(),
  timePlacedFrom: Yup.string()
    .trim()
    .test('duration', 'The start date must be before the end date', function (value) {
      return checkDuration(value, this.parent.timePlacedTo, this);
    })
    .nullable(),
  timePlacedTo: Yup.string().trim().nullable(),
  timeUpdatedFrom: Yup.string()
    .trim()
    .test('duration', 'The start date must be before the end date', function (value) {
      return checkDuration(value, this.parent.timeUpdatedTo, this);
    })
    .nullable(),
  timeUpdatedTo: Yup.string().trim().nullable(),
  subscriptionId: Yup.string().trim(),
});

export const OrderfilterFormWhiteList = [
  'id',
  'orderId',
  'memberId',
  'address',
  'email',
  'name',
  'phone',
  'ipAddress',
  'logonId',
  'status',
  'statuses',
  'excludedStatuses',
  'siteId',
  'paypalEmail',
  'payReferenceIds',
  'account',
  'donationOrgId',
  'subscriptionId',
  'parentOrderId',
  'blockReason',
  'fulfillmentCenter',
  'partNumber',
];

export function isValidDateRange(startDateVal, endDateVal) {
  if (startDateVal && endDateVal) {
    const startDate = new Date(startDateVal);
    const endDate = new Date(endDateVal);
    if (startDate.getTime() > endDate.getTime()) {
      return false;
    }
    const twoWeeksAgo = subWeeks(endDate, 2);
    if (startDate.getTime() < twoWeeksAgo.getTime()) {
      return false;
    }
  }
  return true;
}

const orderFiltersOnlySearchWithThemNotAllowed = new Set(['city', 'zip']);

/**
 * Returns a list of errors on a filter form if there are any
 *
 * @param {object} validationSchema // The schema used to validate the current form
 * @param {object} values // The current values for the form
 * @returns {object} // The list of errors the form currently has.  If none then the form is currently valid or not dirty
 */
export const doFilterFormValidation = async (
  validationSchema,
  filterFormWhiteList = [],
  values,
) => {
  let errors = {};

  if (!values) {
    return errors;
  }

  const validationOptions = {
    abortEarly: false,
  };

  await validationSchema.validate(values, validationOptions).catch((error) => {
    for (let i = 0; i < error.inner.length; i++) {
      const currentError = error.inner[i];
      const currErrorParams = currentError.params;

      if (currErrorParams.value !== '' && currErrorParams.value !== null) {
        errors[currErrorParams.path] = currentError.message;
      }
    }
  });

  if (JSON.stringify(errors) === '{}' && filterFormWhiteList.length > 0) {
    const results = Object.entries(values).filter(
      ([name, value]) => value !== '' && value !== null,
    );

    if (results.length === 1 && !filterFormWhiteList.includes(results[0][0])) {
      errors[results[0][0]] = 'There must be at least 2 items to search on';
    }

    if (results.length === 3 && orderFiltersOnlySearchWithThemNotAllowed.has(results[0][0])) {
      errors[results[0][0]] =
        'at least one of memberId, address, name, phone, statuses, excludedStatuses, siteId, paypalEmail, payReferenceIds, account, donationOrgId, subscriptionId, parentOrderId, blockReasons, fulfillmentCenters, partNumbers, or trackingId must be set';
    }
  }

  return errors;
};

export function checkDuration(startDateStr, endDateStr, ctx) {
  if (startDateStr && endDateStr) {
    const startDate = new Date(startDateStr);
    const endDate = new Date(endDateStr);
    if (!isValid(startDate) || !isValid(endDate)) {
      return false;
    }
    if (startDate.getTime() > endDate.getTime()) {
      return false;
    }
    const twoWeeksAgo = subWeeks(endDate, 2);
    if (startDate.getTime() < twoWeeksAgo.getTime()) {
      return ctx.createError({ message: 'The date range should be less than 2 weeks' });
    }
  }
  return true;
}
