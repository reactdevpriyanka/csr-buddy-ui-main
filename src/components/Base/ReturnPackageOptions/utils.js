export const LABEL_CHOICES = [
  {
    label: '1 Label',
    enabled: true,
    visible: true,
    value: 1,
  },
  {
    label: '2 Labels',
    enabled: true,
    visible: true,
    value: 2,
  },
  {
    label: '3 Labels',
    enabled: true,
    visible: true,
    value: 3,
  },
  {
    label: '4 Labels',
    enabled: true,
    visible: true,
    value: 4,
  },
  {
    label: '5 Labels',
    enabled: true,
    visible: true,
    value: 5,
  },
  {
    label: '6 Labels',
    enabled: true,
    visible: true,
    value: 6,
  },
];

export const BOX_CHOICES = [
  {
    label: '1 Box',
    enabled: true,
    visible: true,
    value: 1,
  },
  {
    label: '2 Boxes',
    enabled: true,
    visible: true,
    value: 2,
  },
  {
    label: '3 Boxes',
    enabled: true,
    visible: true,
    value: 3,
  },
  {
    label: '4 Boxes',
    enabled: true,
    visible: true,
    value: 4,
  },
  {
    label: '5 Boxes',
    enabled: true,
    visible: true,
    value: 5,
  },
  {
    label: '6 Boxes',
    enabled: true,
    visible: true,
    value: 6,
  },
];
