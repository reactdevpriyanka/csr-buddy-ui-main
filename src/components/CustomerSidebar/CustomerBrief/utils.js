export const isWeight = (name) => name === 'WEIGHT_LIMIT';

export const blacklistedTags = new Set(['DISCOUNT_ABUSER', 'RETURNS_ABUSER', 'SUSPECTED_FRAUD']);
