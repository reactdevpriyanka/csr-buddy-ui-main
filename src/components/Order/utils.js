import { formatDeliveredDate, formatDeliveryDate, includesTime } from '@utils/dates';

export const itemGroupingHeader = (shippingStep, isMultipleShipments, ordinal, total) => {
  switch (shippingStep) {
    case 'ORDER_PLACED':
      return 'Processing Order';
    case 'PACKING_ITEMS':
      return 'Packing Order';
    default:
      return isMultipleShipments ? `Shipment ${ordinal} of ${total}` : `Shipment ${ordinal}`;
  }
};

export function getDeliveryEstimate(trackingData, deliveryDescriptions) {
  if (!trackingData) {
    return;
  }

  const isDelivered = trackingData.shippingStep === 'DELIVERED';

  const deliveryDescription = isDelivered ? 'Delivered on:' : 'Estimated Delivery:';
  // : `${deliveryDescriptions[trackingData.shippingStatus]}:`;

  let dateISO = isDelivered
    ? trackingData.trackingEvent?.date
    : trackingData.derivedDeliveryDate || trackingData.estimatedDeliveryDate;

  if (dateISO && !includesTime(dateISO)) {
    dateISO = `${dateISO}`;
  }

  const formattedDeliveryDate = formatDeliveryDate(dateISO, trackingData.shippingStep);

  return {
    deliveryDescription,
    formattedDeliveryDate,
  };
}

export function getDeliveredDeliveryEstimate(trackingData, deliveryDescriptions) {
  if (!trackingData) {
    return;
  }

  const isDelivered = trackingData.shippingStep === 'DELIVERED';

  const deliveryDescription = isDelivered ? 'Delivered on:' : 'Estimated Delivery:';
  // : `${deliveryDescriptions[trackingData.shippingStatus]}:`;

  let dateISO = isDelivered ? trackingData.trackingEvent?.date : trackingData.actualDeliveryTime;

  if (dateISO && !includesTime(dateISO)) {
    dateISO = `${dateISO}`;
  }

  const formattedDeliveryDate = formatDeliveredDate(dateISO, trackingData.shippingStep);

  return {
    deliveryDescription,
    formattedDeliveryDate,
  };
}

export function getLegacyDeliveryEstimate(tracking, deliveryDescriptions) {
  const isDelivered = tracking?.shippingStep === 'DELIVERED';

  const deliveryDescription = isDelivered
    ? 'Delivered on:'
    : `${deliveryDescriptions[tracking?.shippingStatus]}:`;

  const dateISO = isDelivered
    ? tracking?.trackingEvent && tracking?.trackingEvent.date
    : tracking?.estimatedDeliveryDate;

  const formattedDeliveryDate = formatDeliveryDate(dateISO, tracking?.shippingStep);

  return {
    deliveryDescription,
    formattedDeliveryDate,
  };
}

export const ORDER_ATTRIBUTE = {
  CANCELLABLE: 'CANCELLABLE',
  ADDRESS_UPDATABLE: 'ADDRESS_UPDATABLE',
  EDITABLE: 'EDITABLE',
  PRESCRIPTION: 'PRESCRIPTION',
  AUTOSHIP: 'AUTOSHIP',
};

export const LINE_ITEM_ATTRIBUTE = {
  QTY_REDUCIBLE: 'QTY_REDUCIBLE',
  CANCELLABLE: 'CANCELLABLE',
};

export const ORDER_STATUS = {
  CANCELED: 'CANCELED',
  IN_PROGRESS: 'IN_PROGRESS',
  ACTIVE:'ACTIVE'
};

export const ORDER_CANCEL_REASONS = [
  {
    label: 'Customer Initiated',
  },
  {
    label: 'Customer Request',
  },
  {
    label: 'Duplicate Order',
  },
  {
    label: 'Fraud',
  },
  {
    label: 'Gift Card Limit',
  },
  {
    label: 'Inventory Issue',
  },
  {
    label: 'No Response',
  },
  {
    label: 'Other',
  },
  {
    label: 'Payment Issue',
  },
  {
    label: 'Prescription Issue',
  },
  {
    label: 'Reseller',
  },
];
