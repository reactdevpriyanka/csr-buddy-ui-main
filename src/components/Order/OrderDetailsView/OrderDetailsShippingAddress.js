import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import { Link } from '@material-ui/core';
import { Card } from '@mui/material';
import toFormattedPhoneNumber from '@/utils/formatters';
import { useState } from 'react';
import useAthena from '@/hooks/useAthena';
import ShowShippingAddressDialog from './Dialogs/ShowShippingAddressDialog';
import { AllowableActions } from './utils';

const useStyles = makeStyles((theme) => ({
  root: {
    '& li:first-of-type': {
      whiteSpace: 'nowrap',
      display: 'block',
    },
  },
  text: {
    fontFamily: 'Roboto',
    fontWeight: '400',
    fontStyle: 'normal',
    fontSize: '0.875rem',
    lineHeight: '1rem',
    color: '#121212',
    marginBottom: '0.3rem',
  },
  content: {
    paddingLeft: theme.utils.fromPx(16),
    paddingRight: theme.utils.fromPx(16),
    paddingBottom: theme.utils.fromPx(16),
  },

  title: {
    fontFamily: 'Roboto',
    marginTop: theme.utils.fromPx(16),
    marginLeft: '0.9375rem',
    fontWeight: '700',
    fontSize: theme.utils.fromPx(16),
    color: theme.palette.blue.dark,
  },
  title2: {
    fontFamily: 'Roboto',
    color: '#666666',
    fontSize: '0.95rem',
    marginBottom: '1rem',
  },
  shippingTitle2: {
    fontFamily: 'Roboto',
    color: '#666666',
    fontSize: theme.utils.fromPx(16),
    display: 'inline-block',
    justifyContent: 'space-between',
  },
  link: {
    float: 'right',
    display: 'inline',
    fontWeight: 600,
    color: '#1C49C2',
    '&:focus, &:hover': {
      cursor: 'pointer',
    },
  },
  hr: {
    flexGrow: '0',
    border: '1px solid #CCCCCC',
  },
  parentCard: {
    borderRadius: '0.25rem',
  },
  shippingAddress: {
    whiteSpace: 'nowrap',
    marginTop: theme.utils.fromPx(16),
  },
}));

const OrderDetailsShippingAddress = ({
  shippingAddress,
  giftCardOnlyEmail,
  orderNumber,
  isActionAllowed,
}) => {
  const classes = useStyles();

  const [openShippingAddressDialog, setOpenShippingAddressDialog] = useState(false);

  const { getLang } = useAthena(); // athena config

  return (
    <div data-testid="orderDetailsShippingAddress" className={classes.root}>
      <Card className={classes.parentCard} elevation={0}>
        <div className={classes.title}> {'Recipient'} </div> <br />
        <div className={classes.content}>
          <div className={classes.title2}>
            {getLang('shippingRecipient', { fallback: 'Recipient' })}
            <div>
              <span data-testid="customer-shippingAddress-name" className={classes.text}>
                {giftCardOnlyEmail ? giftCardOnlyEmail : shippingAddress?.fullName}
              </span>
              <br />
            </div>
          </div>
          <hr className={classes.hr} />
          <div className={classes.shippingAddress}>
            {!giftCardOnlyEmail && shippingAddress ? (
              <>
                <span className={classes.shippingTitle2}>
                  {getLang('shippingAddress', { fallback: 'Shipping Address:' })}
                </span>
                {isActionAllowed({ actionName: AllowableActions.EDIT_SHIPPING_ADDRESS }) ? (
                  <Link
                    data-testid="orderDetailsShippingAddress:update"
                    className={classes.link}
                    onClick={() => {
                      setOpenShippingAddressDialog(true);
                    }}
                  >
                    Update
                  </Link>
                ) : undefined}
                <br />
                {shippingAddress.addressLine1}
                <br />
                {shippingAddress.addressLine2}
                {shippingAddress.addressLine2 && <br />}
                {shippingAddress.city}, {shippingAddress.state}, {shippingAddress.postcode},{' '}
                {shippingAddress.country}
                <br />
                {'P: '}
                {toFormattedPhoneNumber(shippingAddress.phone)}
                <br />
              </>
            ) : (
              <>
                <span className={classes.shippingTitle2}>
                  {getLang('shippingAddress', { fallback: 'Shipping Address:' })}
                </span>
                <br />
                {'N/A'}
              </>
            )}
          </div>
        </div>
      </Card>
      {openShippingAddressDialog && (
        <ShowShippingAddressDialog
          orderNumber={orderNumber}
          isOpen={openShippingAddressDialog}
          openDialog={setOpenShippingAddressDialog}
          shippingAddress={shippingAddress}
        />
      )}
    </div>
  );
};

OrderDetailsShippingAddress.propTypes = {
  shippingAddress: PropTypes.object,
  giftCardOnlyEmail: PropTypes.string,
  orderNumber: PropTypes.string.isRequired,
  isActionAllowed: PropTypes.func,
};

export default OrderDetailsShippingAddress;
