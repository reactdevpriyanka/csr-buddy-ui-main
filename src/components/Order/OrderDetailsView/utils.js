import { getDayDateYearTimeTimezone } from '@/utils';
import { multiply } from 'lodash';

export const AllowableActions = {
  RETURN_ITEMS: 'RETURN_ITEMS',
  SEND_ORDER_INVOICE: 'SEND_ORDER_INVOICE',
  SEND_ORDER_CONFIRMATION: 'SEND_ORDER_CONFIRMATION',
  SEND_ORDER_CANCELED_NOTIFICATION: 'SEND_ORDER_CANCELED_NOTIFICATION',
  ALLOCATE_INVENTORY: 'ALLOCATE_INVENTORY',
  ADD_FULFILLMENT_COMMENT: 'ADD_FULFILLMENT_COMMENT',
  BLOCK_ORDER: 'BLOCK_ORDER',
  FORCE_RELEASE_ORDER: 'FORCE_RELEASE_ORDER',
  PROCESS_ORDER: 'PROCESS_ORDER',
  CANCEL_ORDER: 'CANCEL_ORDER',
  EDIT_SHIPPING_ADDRESS: 'EDIT_SHIPPING_ADDRESS',
  ADD_PROMOTION_CODE: 'ADD_PROMOTION_CODE',
  ADD_PRICE_ADJUSTMENT: 'ADD_PRICE_ADJUSTMENT',
  REMOVE_ITEM: 'REMOVE_ITEM',
  REDUCE_QUANTITY: 'REDUCE_QUANTITY',
  CREATE_PACKAGE: 'CREATE_PACKAGE',
  CONFIRM_SHIPMENT: 'CONFIRM_SHIPMENT',
  RESEND_SHIPMENT_MESSAGE: 'RESEND_SHIPMENT_MESSAGE',
  FORCE_CANCEL_RELEASE: 'FORCE_CANCEL_RELEASE',
  CANCEL_RELEASE: 'CANCEL_RELEASE',
  CANCEL_RETURN: 'CANCEL_RETURN',
  MARK_AS_RECEIVED: 'MARK_AS_RECEIVED',
  CREATE_NEW_LABELS: 'CREATE_NEW_LABELS',
};

export const AllowableActionTypes = {
  ORDER_ACTIONS: 'orderActions',
  RELEASE_ACTIONS: 'releaseActions',
  RETURN_ACTIONS: 'returnActions',
};

export const PackageReleaseTypes = Object.freeze({
  PACKAGE: Symbol('package'),
  RELEASE: Symbol('release'),
});

// format supplemental data attributes to for tooltip text
export const getAttributes = (tag, attributesMap, vetContactInfo, product, rowData, testId) => {
  switch (tag) {
    case 'DISCONTINUED':
      return (
        <span>
          {product?.discontinueDate != null && (
            <span data-testid="tag:discontinueDate" key="discontinueDate">
              <b>Discontinued Date:</b> {getDayDateYearTimeTimezone(product.discontinueDate)}
            </span>
          )}
        </span>
      );
    // case 'OUT_OF_STOCK': Disabling this until Force BackOrder and OOS back in stock dates are available in the api
    //   return (
    //     <span>
    //       {product?.discontinueDate != null && (
    //         <span key="discontinueDate">
    //           <b>Back in Stock Date:</b>
    //         </span>
    //       )}
    //     </span>
    //   );
    // case 'FORCED_BACK_ORDER':
    //   return (
    //     <span>
    //       {product?.discontinueDate != null && (
    //         <span key="discontinueDate">
    //           <b>Available Date:</b>
    //         </span>
    //       )}
    //     </span>
    //   );
    case 'CONNECT_WITH_A_VET':
      return attributesMap ? (
        <span>
          {vetContactInfo?.petName != null && (
            <span data-testid="tag:petName" key="petName">
              Pet Name: {vetContactInfo.petName}
              <br />
            </span>
          )}
          {attributesMap?.reason != null && (
            <span data-testid="tag:appointmentReason" key="appointmentReason">
              Appointment Reason: {attributesMap.reason}
              <br />
            </span>
          )}
          {attributesMap?.id != null && (
            <span data-testid="tag:appointmentId" key="appointmentId">
              Appointment ID: {attributesMap.id}
              <br />
            </span>
          )}
          {attributesMap?.endTime != null && (
            <span data-testid="tag:appointmentEndTime" key="appointmentEndTime">
              Appointment End Time: {`${getDayDateYearTimeTimezone(attributesMap.endTime)}`}
              <br />
            </span>
          )}
        </span>
      ) : null;
    case 'PHARMACY':
      return vetContactInfo ? (
        <span>
          {vetContactInfo?.clinicName != null && (
            <span data-testid="tag:clinicName" key="clinicName">
              Clinic Name: {vetContactInfo.clinicName}
              <br />
            </span>
          )}
          {vetContactInfo?.contactVet != null && (
            <span data-testid="tag:vetContacted" key="vetContacted">
              Vet Contacted: {vetContactInfo.contactVet ? 'Yes' : 'No'}
              <br />
            </span>
          )}
          {vetContactInfo?.vetProfileId != null && (
            <span data-testid="tag:vetProfileId" key="vetProfileId">
              Vet Profile ID: {vetContactInfo.vetProfileId}
              <br />
            </span>
          )}
          {vetContactInfo?.petName != null && (
            <span data-testid="tag:petName" key="petName">
              Pet Name: {vetContactInfo.petName}
              <br />
            </span>
          )}
        </span>
      ) : null;
    case 'DROPSHIP':
      return (
        <span>
          {rowData?.minDeliveryDate != null && rowData?.maxDeliveryDate != null && (
            <span data-testid="tag:minDeliveryDate" key="minDeliveryDate">
              <b>Delivers between:</b> {getDayDateYearTimeTimezone(rowData.minDeliveryDate)} and{' '}
              {getDayDateYearTimeTimezone(rowData.maxDeliveryDate)}
            </span>
          )}
        </span>
      );
    case 'PERSONALIZED':
      const entries = Object.entries(product?.personalizationAttributeMap || {}).sort(
        (obj1, obj2) => {
          return obj1[0].localeCompare(obj2[0]);
        },
      );
      return (
        <span>
          {entries?.map((entry) => {
            return (
              <div key={entry[0]}>
                <span data-testid={`tag:personalized:${entry[0]}`}>
                  <b>{entry[0]}:</b> {entry[1]}
                </span>
              </div>
            );
          })}
        </span>
      );
    default:
      return null;
  }
};

export const generateReturnItemKey = (returnItem) => returnItem.type + '_' + returnItem.state;

// Sums up the total quantity returned for each type of unique return
export const createTotalReturnedQuantities = (returnItems) => {
  const totalReturnedQuantities = {};

  for (const returnItem of returnItems) {
    const key = returnItem.lineItemId;
    const innerKey = generateReturnItemKey(returnItem);

    if (!totalReturnedQuantities[key]) {
      totalReturnedQuantities[key] = {};
    }

    if (!totalReturnedQuantities[key][innerKey]) {
      totalReturnedQuantities[key][innerKey] = 0;
    }

    // When either type of concession fails, the return service gives us back a totalCredit of 0 and no quantity field.
    // Because of this, we treat each concession like a totalCredit of 1 so it can be distributed to shipments like other concessions
    let specialCaseIncrement = 0;
    if (innerKey === 'PRODUCT_CONCESSION_FAILED' || innerKey === 'SHIPPING_CONCESSION_FAILED') {
      specialCaseIncrement = 1;
    }

    const increment =
      returnItem.type === 'PRODUCT_CONCESSION' || returnItem.type === 'SHIPPING_CONCESSION'
        ? returnItem?.totalCredit + specialCaseIncrement
        : returnItem?.quantity + specialCaseIncrement;

    totalReturnedQuantities[key][innerKey] += increment;
  }

  return totalReturnedQuantities;
};

export const orderHasItemsSpreadAcrossShipments = (shipments, currShipment, itemMap = {}) => {
  if (shipments?.length === 1) return false;

  const lineItemIds = Object.keys(itemMap);
  for (const lineItemId of lineItemIds) {
    const otherShipmentLineItemIds = shipments
      ?.filter((shipment) => shipment.id !== currShipment.id)
      ?.map((shipment) => shipment?.shipmentItems)
      ?.flat()
      ?.map(({ lineItemId }) => lineItemId);
    if (otherShipmentLineItemIds?.includes(lineItemId)) return true;
  }
  return false;
};

// When the same lineItemId is found across multiple shipments, we have no way of knowing which returnItems (if any)
// are attributed to any specific item in any specific shipment.
//
// The number of returns processed for a lineItemId will always be less than the total quantity of that lineItemId across all shipments
// so we iterate through the shipments untill we run out of return tags for that lineItemId.
export const assignReturnsToShipments = (shipments, returnItems, itemMap = {}) => {
  const totalReturnedQuantities = createTotalReturnedQuantities(returnItems);

  for (const [index, shipment] of shipments.entries()) {
    for (const shipmentItem of shipment.shipmentItems) {
      const lineItemId = shipmentItem.lineItemId;

      if (totalReturnedQuantities[lineItemId]) {
        const returnTagsToRender = Object.entries(totalReturnedQuantities[lineItemId])
          ?.filter(([returnItemKey, quantityReturned]) => quantityReturned >= 0.001)
          ?.map(([returnItemKey, quantityReturned]) => returnItemKey);

        shipments[index] = {
          ...shipment,
          returnTagsToRender: returnTagsToRender,
        };

        for (const returnItemKey in totalReturnedQuantities[lineItemId]) {
          const returnedQuantity = totalReturnedQuantities[lineItemId][returnItemKey];
          if (returnedQuantity < 0) continue; // No need to continue to compute if there is no returnable quantity left for a lineItemId

          const returnItem = returnItems.find(
            (item) => generateReturnItemKey(item) === returnItemKey,
          );

          if (!returnItem) continue;

          let quantityToDecrement;

          switch (returnItem?.type) {
            case 'PRODUCT_CONCESSION':
              quantityToDecrement = multiply(
                shipmentItem?.quantity * Number.parseFloat(itemMap[lineItemId]?.unitPrice?.value),
              );
              break;
            case 'SHIPPING_CONCESSION':
              quantityToDecrement = returnItem?.totalCredit;
              break;
            case 'REPLACEMENT':
            case 'REFUND':
              quantityToDecrement = shipmentItem?.quantity;
              break;
            default:
              quantityToDecrement = 0;
              break;
          }

          // Subtract the decrement quantity from the quantity of items (or total value of products in concession scenarios)
          // returned to only add returns to be rendered to a shipment if there were enough returns to justify it
          // when the same lineItemId is spread across multiple packages
          totalReturnedQuantities[lineItemId][returnItemKey] =
            returnedQuantity - quantityToDecrement;
        }
      }
    }
  }
};
