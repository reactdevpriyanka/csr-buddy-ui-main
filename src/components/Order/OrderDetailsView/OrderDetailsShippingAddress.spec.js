import { renderWrap } from '@/utils';
import orderDetailsViewData from '__mock__/orderdetails/orderdetailsview';
import OrderDetailsShippingAddress from './OrderDetailsShippingAddress';

describe('<OrderDetailsShippingAddress />', () => {
  const defaultProps = {
    ...orderDetailsViewData,
    isActionAllowed: () => true,
  };

  const render = renderWrap(OrderDetailsShippingAddress, { defaultProps });

  test('it should render children', () => {
    expect(render()).toMatchSnapshot();
  });

  test('it should render update address link', () => {
    const { getByTestId } = render();
    expect(getByTestId(`orderDetailsShippingAddress:update`)).toBeTruthy();
  });

  describe('when isActionAllowed equals false', () => {
    const defaultProps = {
      ...orderDetailsViewData,
      isActionAllowed: () => false,
    };

    const render = renderWrap(OrderDetailsShippingAddress, { defaultProps });

    test('it should not render update address link', () => {
      const { queryByTestId } = render();
      expect(queryByTestId(`orderDetailsShippingAddress:update`)).not.toBeInTheDocument();
    });
  });
});
