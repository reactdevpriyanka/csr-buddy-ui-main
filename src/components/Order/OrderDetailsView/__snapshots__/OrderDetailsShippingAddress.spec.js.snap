// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderDetailsShippingAddress /> it should render children 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-1"
        data-testid="orderDetailsShippingAddress"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root makeStyles-parentCard-9 css-xg2nla-MuiPaper-root-MuiCard-root"
        >
          <div
            class="makeStyles-title-4"
          >
             
            Recipient
             
          </div>
           
          <br />
          <div
            class="makeStyles-content-3"
          >
            <div
              class="makeStyles-title2-5"
            >
              Recipient
              <div>
                <span
                  class="makeStyles-text-2"
                  data-testid="customer-shippingAddress-name"
                >
                  Erica Deleoz
                </span>
                <br />
              </div>
            </div>
            <hr
              class="makeStyles-hr-8"
            />
            <div
              class="makeStyles-shippingAddress-10"
            >
              <span
                class="makeStyles-shippingTitle2-6"
              >
                Shipping Address:
              </span>
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-link-7 MuiTypography-colorPrimary"
                data-testid="orderDetailsShippingAddress:update"
              >
                Update
              </a>
              <br />
              1950 N STEMMONS FWY
              <br />
              STE 3000
              <br />
              DALLAS
              , 
              TX
              , 
              75207-3203
              ,
               
              US
              <br />
              P: 
              480-524-7058
              <br />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-1"
      data-testid="orderDetailsShippingAddress"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root makeStyles-parentCard-9 css-xg2nla-MuiPaper-root-MuiCard-root"
      >
        <div
          class="makeStyles-title-4"
        >
           
          Recipient
           
        </div>
         
        <br />
        <div
          class="makeStyles-content-3"
        >
          <div
            class="makeStyles-title2-5"
          >
            Recipient
            <div>
              <span
                class="makeStyles-text-2"
                data-testid="customer-shippingAddress-name"
              >
                Erica Deleoz
              </span>
              <br />
            </div>
          </div>
          <hr
            class="makeStyles-hr-8"
          />
          <div
            class="makeStyles-shippingAddress-10"
          >
            <span
              class="makeStyles-shippingTitle2-6"
            >
              Shipping Address:
            </span>
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-link-7 MuiTypography-colorPrimary"
              data-testid="orderDetailsShippingAddress:update"
            >
              Update
            </a>
            <br />
            1950 N STEMMONS FWY
            <br />
            STE 3000
            <br />
            DALLAS
            , 
            TX
            , 
            75207-3203
            ,
             
            US
            <br />
            P: 
            480-524-7058
            <br />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
